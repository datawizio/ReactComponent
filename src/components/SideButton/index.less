@import "../Button/index.less";

@padding: 15px;
@iconMargin: 13px;
@sideMargin: @iconMargin * 2 + @padding;

.side-button {
  position: fixed !important;

  &:hover {
    transform: translateX(0);
  }
}

.side-button--right {
  right: 0;
  padding: 0 @padding 0 0 !important;
  transform: translateX(calc(100% - @sideMargin));

  .side-button__icon {
    margin: 0 @iconMargin;
  }
}

.side-button--left {
  direction: ltr;
  left: 0;
  padding: 0 0 0 @padding !important;
  transform: translateX(calc(-1 * (100% - @sideMargin)));

  .side-button__icon {
    margin: 0 @iconMargin;
  }
}
