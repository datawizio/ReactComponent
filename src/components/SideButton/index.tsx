import * as React from "react";
import { createPortal } from "react-dom";

import Button from "../Button";
import { ButtonProps } from "antd/lib/button";

import "./index.less";

export interface SideButtonProps extends ButtonProps {
  text: string;
  className: string;
  renderTo?: Element;
  icon: React.ReactNode;
  side?: "left" | "right";
  extra?: React.ReactNode;
}

const SideButton: React.FC<SideButtonProps> = props => {
  const {
    className,
    icon,
    text,
    side,
    renderTo,
    style,
    extra,
    ...restProps
  } = props;

  return createPortal(
    <div
      style={style}
      className={`side-button-container side-button-container--${side} ${className}`}
    >
      <Button {...restProps} className={"side-button side-button--" + side}>
        {side === "right" && (
          <>
            <span className="side-button__icon">{icon}</span>
            <span className="side-button__text">{text}</span>
          </>
        )}

        {side === "left" && (
          <>
            <span className="side-button__text">{text}</span>
            <span className="side-button__icon">{icon}</span>
          </>
        )}
      </Button>

      {extra}
    </div>,
    renderTo
  );
};

SideButton.defaultProps = {
  side: "right",
  renderTo: window.document.body
};

export default SideButton;
