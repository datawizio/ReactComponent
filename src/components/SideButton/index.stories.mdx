import { Meta, Story, ArgsTable } from "@storybook/addon-docs/blocks";
import { SearchOutlined } from "@ant-design/icons";

import SideButton from "./index";

<Meta title="Components/Buttons/SideButton" component={SideButton} />

# SideButton

The `SideButton` component is used to create buttons that are positioned on the sides.

<Story name="Basic">
  <div>
    <SideButton
      className="basic-side-button"
      style={{ top: 50 }}
      type="primary"
      icon={<SearchOutlined />}
      text="primary"
    />
    <SideButton
      className="basic-side-button"
      style={{ top: 100 }}
      type="primary"
      icon={<SearchOutlined />}
      text="ghost"
      ghost
    />
    <SideButton
      className="basic-side-button"
      style={{ top: 150 }}
      type="default"
      icon={<SearchOutlined />}
      text="default"
    />
    <SideButton
      className="basic-side-button"
      style={{ top: 200 }}
      type="dashed"
      icon={<SearchOutlined />}
      text="dashed"
    />
  </div>
</Story>


<Story name="Small">
  <div>
    <SideButton
      className="small-side-button"
      size="small"
      style={{ top: 300 }}
      type="primary"
      icon={<SearchOutlined />}
      text="primary"
    />
    <SideButton
      className="small-side-button"
      size="small"
      style={{ top: 350 }}
      type="primary"
      icon={<SearchOutlined />}
      text="ghost"
      ghost
    />
    <SideButton
      className="small-side-button"
      size="small"
      style={{ top: 400 }}
      type="default"
      icon={<SearchOutlined />}
      text="default"
    />
    <SideButton
      className="small-side-button"
      size="small"
      style={{ top: 450 }}
      type="dashed"
      icon={<SearchOutlined />}
      text="dashed"
    />
  </div>
</Story>

## Props

<ArgsTable of={SideButton} />
