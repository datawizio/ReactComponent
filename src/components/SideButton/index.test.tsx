import "jsdom-global/register";
import "@testing-library/jest-dom";

import React, { ReactElement } from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";

import SideButton, { SideButtonProps } from "./index";

const mockProps: SideButtonProps = {
  text: "Test Button",
  className: "test-class",
  icon: <span>Icon</span>,
  side: "right"
};

const setUp = (props: Partial<SideButtonProps> = {}) => {
  return shallow(<SideButton {...mockProps} {...props} />);
};

describe("SideButton component", () => {
  let component: ReturnType<typeof setUp>;

  beforeEach(() => {
    component = setUp();
  });

  it("is rendered correctly", () => {
    expect(component.exists()).toBeTruthy();
  });

  it("matches snapshot", () => {
    expect(component.debug()).toMatchSnapshot();
  });

  it("has the correct class for the 'right' side", () => {
    expect(component.find(".side-button--right").length).toBeTruthy();
  });

  it("has the correct class for the 'left' side", () => {
    const wrapper = setUp({ side: "left" });
    expect(wrapper.find(".side-button--left").length).toBeTruthy();
  });

  it("renders text correctly", () => {
    expect(component.find(".side-button__text").text()).toBe(mockProps.text);
  });

  it("renders icon correctly", () => {
    expect(
      component
        .find(".side-button__icon")
        .contains(mockProps.icon as ReactElement)
    ).toBeTruthy();
  });

  it("applies custom className", () => {
    expect(component.find(".test-class").length).toBeTruthy();
  });

  it("is rendered on the default side when no side is provided", () => {
    const wrapper = setUp({ side: undefined });
    expect(wrapper.find(".side-button--right").length).toBeTruthy();
  });

  it("renders extra content when provided", () => {
    const extraContent = <div className="extra-content">Extra</div>;
    const wrapper = setUp({ extra: extraContent });
    expect(wrapper.contains(extraContent)).toBeTruthy();
  });

  it("is rendered in the specified container using `createPortal`", () => {
    const customContainer = document.createElement("div");
    const portalMock = jest
      .spyOn(ReactDOM, "createPortal")
      .mockImplementation(node => node);

    setUp({ renderTo: customContainer });
    expect(portalMock).toHaveBeenCalledWith(expect.anything(), customContainer);

    portalMock.mockRestore();
  });
});
