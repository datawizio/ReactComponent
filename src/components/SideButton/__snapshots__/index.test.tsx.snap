// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideButton component matches snapshot 1`] = `
"<Portal containerInfo={{...}}>
  <div style={[undefined]} className=\\"side-button-container side-button-container--right test-class\\">
    <Button className=\\"side-button side-button--right\\" border={true}>
      <span className=\\"side-button__icon\\">
        <span>
          Icon
        </span>
      </span>
      <span className=\\"side-button__text\\">
        Test Button
      </span>
    </Button>
  </div>
</Portal>"
`;
